Mailnag/plugins/libnotifyplugin.py

```diff
@@ -255,13 +273,22 @@ def _notify_summary(self, new_mails, all_mails):
 		self._notifications['0'].show()
 	
 	
-	def _notify_single(self, mails):
+	def _notify_single(self, new_mails, all_mails):
+		# Remove notifications for messages not in all_mails:
+		for k, n in list(self._notifications.items()):
+			if hasattr(n, 'mail') and not (n.mail in all_mails):
+				# The user may have closed the notification:
+				try_close(n)
+				del self._notifications[k]
+
 		# In single notification mode new mails are
 		# added to the *bottom* of the notification list.
-		mails.sort(key = lambda m: m.datetime, reverse = False)
+		new_mails.sort(key = lambda m: m.datetime, reverse = False)
 		
-		for mail in mails:
+		for mail in new_mails:
 			n = self._get_notification(self._get_sender(mail), mail.subject, "mail-unread")
+			# Remember the associated message, so we know when to remove the notification:
+			n.mail = mail
```

Comment:
One optimization request: Could you please store just the mail.id (md5sum) instead of the whole mail object? This would improve both, speed and mem consumption :)

Reply:
Unless I'm confused, that line is just storing a reference to the mail object, not the object itself, so it should be fast and not add to the memory consumption.  (The notification should be closed as soon as we learn that the message is gone, and once it is closed, the mail object should be garbage collected.)  Storing just the mail id would make the line that deletes the notifications a bit more complicated, but I can do it if you like.  I don't know how equality of mail objects works, but it could be every-so-slightly faster to compare their ids instead of the full objects.  Seems unlikely to be worth the extra complexity.

I did some tests where I created and deleted 10 messages per second for a while.  I'm not 100% sure, but it does appear that there might be a memory leak somewhere in the code.  But the memory usage grows even if I change _notify_single to be a noop, so if there is a leak, I think it is elsewhere.  VSZ stays constant but RSS grows slowly, so maybe this isn't a leak, but just the nature of how the memory management works?

Reply:
> I don't know how equality of mail objects works, but it could be every-so-slightly faster to compare their ids instead of the full objects.

I believe these are objects of type Mail defined in mails.py.  Since that class doesn't define __eq__, I believe objects are compared by comparing pointers (`id(object)` in python, not to be confused with `object.id`), so it will be fast.  So I don't think there are any optimizations possible here.